### PR TITLE
[FG:InPlacePodVerticalScaling] Disallow removing requests & limits for Burstable pods.

### LINKED ISSUE #128677: [FG:InPlacePodVerticalScaling] Disallow removing requsets & limits during resize
/kind feature

For v1.32 we've decided to remove support for removing requests and limits during a resize. We will revisit this in a future release (https://github.com/kubernetes/kubernetes/issues/128675)

/cc @dchen1107 

/sig node
/priority important-soon

### PR DESCRIPTION
**Pull request by AnishShah**
#### What this PR does / why we need it:

We decided to remove support for removing requests and limits during a resize. We will revisit this in a future release 

#### Which issue(s) this PR fixes:

Fixes #128677

#### Does this PR introduce a user-facing change?

```release-note
Removed support for removing requests and limits during a pod resize.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go]:
```suggestion
			if len(oldResourceList) > len(resourceList) {
```

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5509]: nit: move this out of the loop

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5506]:
```suggestion
	// Do not allow removing resource requests/limits on resize.
	// TODO(#128675): Add Kubelet support for removing resources, and remove this validation check.
```

File: pkg/apis/core/validation/validation.go
Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Done.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5513]: Probably good to give a more precise field path for this. At least separate the requests & limits cases.

File: pkg/apis/core/validation/validation.go
Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Done.

File: pkg/apis/core/validation/validation.go
Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Done.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [reply pkg/apis/core/validation/validation.go]: Wait, I think we still need the per-resource checks. Otherwise, you could remove a resource and add a different one in the same request.

File: pkg/apis/core/validation/validation.go
Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Done.

File: pkg/apis/core/validation/validation.go
Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Added it back.

File: pkg/apis/core/validation/validation.go
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:5512]:
```suggestion
	for ix, ctr := range oldPod.Spec.Containers {
		if resourcesRemoved(newPod.Spec.Containers[ix].Resources.Requests, ctr.Resources.Requests) {
			allErrs = append(allErrs, field.Forbidden(specPath.Child("containers").Index(ix).Child("resources").Child("requests"), "resource requests cannot be removed"))
		}
		if resourcesRemoved(newPod.Spec.Containers[ix].Resources.Limits, ctr.Resources.Limits) {
			allErrs = append(allErrs, field.Forbidden(specPath.Child("containers").Index(ix).Child("resources").Child("limits"), "resource limits cannot be removed"))
		}
	}
```

Comment (tallclair) [reply pkg/apis/core/validation/validation.go:5512]: More precise field errors

Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Done.

File: pkg/apis/core/validation/validation.go
Comment (thockin) [inline pkg/apis/core/validation/validation.go:5507]: When will this TODO be done?  Why do we need to merge this without it?

**Review by thockin — CHANGES_REQUESTED** [review]

File: pkg/apis/core/validation/validation.go
Comment (AnishShah) [reply pkg/apis/core/validation/validation.go:5507]: sidecar support is being added in https://github.com/kubernetes/kubernetes/pull/128367

Comment (tallclair) [reply pkg/apis/core/validation/validation.go:5507]: This PR is racing with https://github.com/kubernetes/kubernetes/pull/128367, but this will probably merge first.

Alternatively, we could just add the sidecar check here, since it's just a no-op without https://github.com/kubernetes/kubernetes/pull/128367

Comment (AnishShah) [reply pkg/apis/core/validation/validation.go]: Added sidecar support

**Comment by wendy-ha18** [comment]
Hi everyone
A gentle reminder that the code freeze has started 02:00 UTC Friday November 8th 2024 . I saw this PR has both `lgtm` and `approved` label already, if you still aim for 1.32 milestone and need time, consider file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) ASAP.

Thanks everyone!!

**Comment by tallclair** [comment]
See https://github.com/kubernetes/kubernetes/pull/128694 for failing resize tests

**Comment by AnishShah** [comment]
The e2e tests added in this PR passed.

**Pull request merged**
